jlog.c

```diff
@@ -1375,17 +1539,69 @@ int jlog_ctx_write_message(jlog_ctx *ctx, jlog_message *mess, struct timeval *wh
     hdr.tv_sec = now.tv_sec;
     hdr.tv_usec = now.tv_usec;
   }
+
+  /* we store the original message size in the header */
   hdr.mlen = mess->mess_len;
-  if (!jlog_file_pwrite(ctx->data, &hdr, sizeof(hdr), current_offset)) {
-    FASSERT(0, "jlog_file_pwrite failed in jlog_ctx_write_message");
-    SYS_FAIL(JLOG_ERR_FILE_WRITE);
+
+  struct iovec v[2];
+  v[0].iov_base = (void *) &hdr;
+  v[0].iov_len = hdr_size;
+
+  if (IS_COMPRESS_MAGIC(ctx)) {
+    if (jlog_compress(mess->mess, mess->mess_len, &v[1].iov_base, (size_t *)&hdr.compressed_len) != 0) {
```

Comment:
You have an upper-bound on the compression size... and as it will likely comfortable fit on stack (say 32k or so) you could use an on stack iov_base if the compression upper bound fits... malloc()ing if it will not.  Then conditionally free later.  You'll avoid ~ 100% of these free/mallocs that way.